fuzzers/libafl-fuzz/src/fuzzer.rs

```diff
@@ -0,0 +1,361 @@
+
+        // Run our fuzzer; WITH CmpLog
+        run_fuzzer_with_stage!(
+            &opt,
+            fuzzer,
+            &mut stages,
+            &mut executor,
+            &mut state,
+            &mut restarting_mgr
+        );
+    } else {
+        // The order of the stages matter!
+        let mut stages = tuple_list!(calibration, power, afl_stats_stage);
+
+        // Run our fuzzer; NO CmpLog
+        run_fuzzer_with_stage!(
+            &opt,
+            fuzzer,
+            &mut stages,
+            &mut executor,
+            &mut state,
+            &mut restarting_mgr
+        );
+    }
+    Ok(())
+    // TODO: serialize state when exiting.
+}
+
+fn base_executor<'a>(
+    opt: &'a Opt,
+    shmem_provider: &'a mut UnixShMemProvider,
+) -> ForkserverExecutorBuilder<'a, UnixShMemProvider> {
+    let mut executor = ForkserverExecutor::builder()
+        .program(opt.executable.clone())
+        .shmem_provider(shmem_provider)
+        .coverage_map_size(opt.map_size)
+        .kill_signal(opt.kill_signal)
+        .debug_child(opt.debug_child)
+        .is_persistent(opt.is_persistent)
+        .is_deferred_frksrv(opt.defer_forkserver)
+        .min_input_size(opt.min_input_len.unwrap_or(AFL_DEFAULT_INPUT_LEN_MIN))
+        .max_input_size(opt.max_input_len.unwrap_or(AFL_DEFAULT_INPUT_LEN_MAX))
+        .timeout(Duration::from_millis(opt.hang_timeout));
+    if let Some(target_env) = &opt.target_env {
+        executor = executor.envs(target_env);
+    }
+    if let Some(harness_input_type) = &opt.harness_input_stdin {
+        executor = executor.parse_afl_cmdline([harness_input_type]);
+    }
+    executor
+}
+
+pub fn fuzzer_target_mode(opt: &Opt) -> Cow<'static, str> {
+    let mut res = String::new();
+    if opt.unicorn_mode {
+        res = format!("{res}unicorn ");
+    }
+    if opt.qemu_mode {
+        res = format!("{res}qemu ");
+    }
+    if opt.forkserver_cs {
+        res = format!("{res}coresight ");
+    }
+    if opt.no_forkserver {
+        res = format!("{res}no_fsrv ");
+    }
+    if opt.crash_mode {
+        res = format!("{res}crash ");
+    }
+    if opt.is_persistent {
+        res = format!("{res}persistent ");
+    }
+    // TODO: not always shmem_testcase
+    res = format!("{res}shmem_testcase ");
+    if opt.defer_forkserver {
+        res = format!("{res}deferred ");
+    }
+    if !(opt.unicorn_mode
+        || opt.qemu_mode
+        || opt.forkserver_cs
+        || opt.non_instrumented_mode
+        || opt.no_forkserver
+        || opt.crash_mode
+        || opt.is_persistent
+        || opt.defer_forkserver)
+    {
+        res = format!("{res}default");
+    }
+    Cow::Owned(res)
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct IsInitialCorpusEntryMetadata {}
+libafl_bolts::impl_serdeany!(IsInitialCorpusEntryMetadata);
+
+pub type LibaflFuzzState =
+    StdState<BytesInput, OnDiskCorpus<BytesInput>, StdRand, OnDiskCorpus<BytesInput>>;
```

Comment:
You should use a CachedOnDiskCorpus for the corpus btw

Reply:
But _in general_ you could keep this generic so we can later slot in more features like Grimoire without too many changes

Reply:
I think the state is fine if it's hard coded since this will probably never change right? ie. the addition of Grimoire should not affect the state type right?

Reply:
The addition of Grimoire etc. will mean the types get more complex and might be easier to let the compiler figure it out. Also copy&paste to other fuzzers may be easier

Reply:
But not necessary to do it either